src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -2372,12 +2402,71 @@ Node *PhiNode::Ideal(PhaseGVN *phase, bool can_reshape) {
       if (ii->is_MergeMem()) {
         MergeMemNode* n = ii->as_MergeMem();
         merge_width = MAX2(merge_width, n->req());
-        saw_self = saw_self || (n->base_memory() == this);
+        if (n->base_memory() == this) {
+          split_must_terminate = true;
+        }
       }
     }
 
-    // This restriction is temporarily necessary to ensure termination:
-    if (!saw_self && adr_type() == TypePtr::BOTTOM)  merge_width = 0;
+    // There are cases with circular dependencies between bottom Phis
+    // and MergeMems. Below is a minimal example.
+    //
+    //               +------------+
+    //               |            |
+    // (base_memory) v            |
+    //              MergeMem      |
+    //                 |          |
+    //                 v          |
+    //                Phi (this)  |
+    //                 |          |
+    //                 v          |
+    //                Phi         |
+    //                 |          |
+    //                 +----------+
+    //
+    // Here, we cannot break the circularity through a self-loop as there
+    // are two Phis involved. Repeatedly splitting the Phis through the
+    // MergeMem leads to non-termination. We check for non-termination below.
+    // Only check for non-termination if necessary.
+    if (!split_must_terminate && adr_type() == TypePtr::BOTTOM &&
+        merge_width > Compile::AliasIdxRaw) {
+      ResourceMark rm;
+      VectorSet visited;
+      Node_List worklist;
+      worklist.push(this);
+      visited.set(this->_idx);
+      auto maybe_add_to_worklist = [&](Node* input) {
+        if (input != nullptr &&
+            (input->is_MergeMem() || input->is_memory_phi()) &&
+            !visited.test_set(input->_idx)) {
+          worklist.push(input);
+          assert(input->adr_type() == TypePtr::BOTTOM,
+                 "should only visit bottom memory");
+        }
+      };
+      split_must_terminate = true; // Assume no circularity until proven otherwise.
+      while (split_must_terminate && worklist.size() > 0) {
```

Comment:
Seems like you have `split_must_terminate` in this condition because the `break` in the `else` path does not exit both loops. When using a separate method, you could directly return false when finding that `split_must_terminate` is false. Then you can remove `split_must_terminate` from this exit condition.

Reply:
Yes, extracting the termination check to a separate function makes it possible to simplify like you suggest. Thanks!